### ISSUE TITLE
Any reason why EnumConverterFactory's target score is MAX_VALUE?

### ISSUE THREAD

**Issue by dicksonleong** [arnaudroger/SimpleFlatMapper#618]
I'm talking about this line: https://github.com/arnaudroger/SimpleFlatMapper/blob/e8f193ab649604163023e8fcdb636d00c7fdfa4f/sfm-converter/src/main/java/org/simpleflatmapper/converter/impl/JavaBaseConverterFactoryProducer.java#L183

My use case: I have a lot of enums that will store as int in database but I don't want to use `NumberToEnumConverter` because I want to specify the int value explicitly, so i create an interface and have all these enums extends from it and SFM will automatically convert them:

```
interface IntPersistable {
    val persistValue: Int
}
```

```
class IntPersistableConverterFactoryProducer : AbstractContextualConverterFactoryProducer() {

    override fun produce(consumer: Consumer<in ContextualConverterFactory<*, *>>?) {
        factoryConverter<Number, IntPersistable>(consumer, object : AbstractContextualConverterFactory<Number, IntPersistable>(Number::class.java, IntPersistable::class.java) {
            override fun newConverter(targetedTypes: ConvertingTypes, contextFactoryBuilder: ContextFactoryBuilder,
                                      vararg params: Any): ContextualConverter<in Number, out IntPersistable> {
                @Suppress("UNCHECKED_CAST")
                return IntPersistableConverter(targetedTypes.to as Class<IntPersistable>)
            }
        })
    }
}
```

However this doesn't work because the EnumConverterFactory's score is higher than my factory. So as a workaround I have to override the score to return the MAX_VALUE on source too (which is a dirty workaround for no reason..?)

```
override fun score(targetedTypes: ConvertingTypes): ConvertingScore {
    return if (TypeHelper.isAssignable(IntPersistable::class.java, targetedTypes.to)) {
        ConvertingScore(ConvertingScore.MAX_SCORE, ConvertingScore.MAX_SCORE)
    } else {
        super.score(targetedTypes)
    }
}
```
The `NumberToEnumConverter` is a convenient generic class so it shouldn't have max score that override user's other implementation class right? 

My second issue is that I don't understand why ConvertingTypes when calculating the target score it calls the `TypeHelper.isAssignable(target, from)`? Shouldn't it be the other way around as `TypeHelper.isAssignable(from, target)` as it need to check if the targeted type can assign to the specified class/interface? (Or i'm using the whole thing wrongly?)
https://github.com/arnaudroger/SimpleFlatMapper/blob/d54829134b008c3a341ceabdd7b51cbeeb4b6787/sfm-converter/src/main/java/org/simpleflatmapper/converter/ConvertingTypes.java#L72

PS. Sorry for Kotlin, i'm too lazy to convert to Java but you should get the idea

**Comment by arnaudroger** [comment]
Very good points... I’ll come back on it later. Enum with special instantiation rules is something that I’ve see a few time. I’ll ha e a deeper look when I can. Thanks for the feedback!

**Comment by arnaudroger** [comment]
Pa the work around would be to add a Custom getter on the column see https://simpleflatmapper.org/0102-getting-started-jdbc.html

**Comment by dicksonleong** [comment]
> Pa the work around would be to add a Custom getter on the column see https://simpleflatmapper.org/0102-getting-started-jdbc.html

But then I have to specify the getter for every enum's mapper right? I don't want to do that (i have too much enums) so I'm looking for a way to specify the getter globally based on a interface that's why i ventured to AbstractContextualConverterFactoryProducer (spent me 1 whole day to understand how to use it)

**Comment by arnaudroger** [comment]
what kind of mapper or you trying to instantiate - jdbc ?
also the emum instantiation is it a static method on the Enum?

**Comment by arnaudroger** [comment]
also just checked I need a refresher the MAX_SCORE is 256, have you try returning a higher score for yours? the score is an int. I'm pretty sure there is nothing stopping it from being bigger.

**Comment by dicksonleong** [comment]
> what kind of mapper or you trying to instantiate - jdbc ?
> also the emum instantiation is it a static method on the Enum?

I'm using JdbcTemplateMapperFactory's RowMapper/ResultSetExtractor.
My coverter is something like this
```kotlin
override fun convert(value: Number?, context: Context): IntPersistable? {
    if (value == null)
        return null

    return targetClass.enumConstants.find { e -> e.persistValue == value }
        ?: throw IllegalArgumentException()
}
```
Basically it access the class's `enumConstants` and loop through it to find the matching enum value. That's why I need to use a factory converter because I need to access to the enum class.

**Comment by arnaudroger** [comment]
have you tried setting your score to 512?

**Comment by arnaudroger** [comment]
so I'm looking into the code and your example so for the targetScore
it's looking if it can assign the target type from the type defined by the factory.
your factory is declared as building IntPersistable which can be a the enum you want or something totally different. so it's not compatible.
I'm trying to build an example to show how it can be done, you can't use the default scoring mechanism there

**Comment by arnaudroger** [comment]
so in the  AbstractContextualConverterFactory you need to override the score function as
```
            @Override
            public ConvertingScore score(ConvertingTypes targetedTypes) {
                if (TypeHelper.isAssignable(IP.class, targetedTypes.getTo())) {
                    return new ConvertingScore(ConvertingTypes.getSourceScore(getFromType(), targetedTypes.getFrom()), ConvertingScore.MAX_SCORE * 2);
                }
                return super.score(targetedTypes);
            }
```

see https://github.com/arnaudroger/SimpleFlatMapper/blob/master/sfm-converter/src/test/java/org/simpleflatmapper/converter/test/FooConverterFactoryProducer.java#L29
and test 
https://github.com/arnaudroger/SimpleFlatMapper/blob/master/sfm-converter/src/test/java/org/simpleflatmapper/converter/test/ConverterServiceTest.java#L184

does that help?

**Comment by dicksonleong** [comment]
yeah, it works.. but still kinda weird have to use max score * 2...

> it's looking if it can assign the target type from the type defined by the factory.
> your factory is declared as building IntPersistable which can be a the enum you want or something totally different. so it's not compatible.

Can you elaborate? Did I use the factory wrongly? Basically IntPersistable is a base interface and all the enums that I wished to convert will be implement this interface. So the target type will be the subclass of the type defined by the factory (means target type assignable to factory defined type), but I don't understand why the default scoring is checking the other way around (is factory defined type assignable to target type)

**Comment by arnaudroger** [comment]
yes it's a bit akward ...

so lets say I create a converter from CharSequence to Object

would it be compatible with Date? or Enum? currently the code says now by
default in the score method.

in your case you now the IntPersistable is compatible with your enum, but
that can't be extrapolated from the type system, it's a feature of your
convention.

**Comment by dicksonleong** [comment]
Well I guess no better way beside rewrite the JavaBaseConverterFactoryProducer to be more extensible, feel free to close this issue.

As a side note: I also don't like the ServiceLoader-based loading on factory producer. A better way will be specify additional factory/factory producer though a global, reuseable instance and mapper can generate from this global instance (eg. Jackson's `ObjectMapper`)

**Comment by arnaudroger** [comment]
well idealy I don't even think the converter is the right place to go,
because it's non prescriptive.

What I really think is it should be easier to register ResultSet Getter for
a type. and eaiser to share that across mapper factories.

**Comment by dicksonleong** [comment]
> What I really think is it should be easier to register ResultSet Getter for a type. and eaiser to share that across mapper factories.

Agreed, I don't want set it everytime creating a mapper for each type that extends the same interface/superclass.. again, this can be solve by using a global shareable factory -based API. Performance wise, it also helps abit since the factory only need to initialize and configure once. But of course this involve a complete rewrite of the current SFM API ...

**Comment by arnaudroger** [comment]
well the mapper are reusable so you would only save on the first time
really.

**Comment by arnaudroger** [comment]
@dicksonleong could you have a look at https://github.com/arnaudroger/SimpleFlatMapper/commit/d19ef84f96d6947305ef69d301fc9a49cbb01e1b#diff-9a042c5271d54a8c2e621b6397ffe746R252 it's java6 style code so a bit verbose but that's what I'm thinking now to make it easier for that king cases

1 - we create a factory seed with getterForType 
2 - we create a newInstance from the configurated seed.

in java8 that would look like that 
```
		JdbcTemplateMapperFactory config = JdbcTemplateMapperFactory.newInstance().addGetterForType(
				type -> TypeHelper.isAssignable(LookupValue.class, type),
				(target, key, mappingContextFactoryBuilder,  properties) -> {
						Enum[] values = EnumHelper.getValues(TypeHelper.toClass(target));
						int index = key.getIndex();
						return (resultSet, context) -> {
							int i = resultSet.getInt(index);
							for(Enum e : values) {
								if (((LookupValue)e).lookupValue() == i) {
									return e;
								}
							}
							return null;
						};
					}
		);
```

**Comment by dicksonleong** [comment]
Interesting.. .this will be available in 6.6.0?

**Comment by arnaudroger** [comment]
Yes it is , 6.6.0 is in maven central
